### ISSUE TITLE
Lib ical allow non compliance with RFC 5545

### ISSUE THREAD

**Issue by ad1rie1** [libical/libical#370]
Hello, 

I repport a bug with libcal serrialisation that allow  RRULE contain both UNTIL and COUNT that is not allowed by RFC 5545. 

see bug : https://bugs.kde.org/show_bug.cgi?id=405884

And the reply from kde team : 
 it appears that this is in fact a bug in libical, not in KCalCore - one 
part is that the serialization code in libical allows this error to happen, 
another part I suspect is deeper in the code that deals with fetching timezone 
information about daylight saving that somehow ends up with both UNTIL and 
COUNT fields having a non-null value.

I suggest you report the issue to libical upstream here: 

        https://github.com/libical/libical

Regards,
Daniel

**Comment by danvratil** [comment]
I can reproduce it when creating an event for Europe/Prague timezone as well, one of the daylight saving rules is broken, too:

```
BEGIN:VTIMEZONE
...
BEGIN:STANDARD
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
DTSTART:19790930T030000
RRULE:FREQ=YEARLY;UNTIL=19961027T030000;COUNT=17;BYDAY=-1SU;BYMONTH=9
RDATE:19950924T030000
END:STANDARD
END:VTIMEZONE
```

This is with libical-3.0.4-1.fc29.x86_64 and kcalcore from master.

**Comment by ad1rie1** [comment]
Hello danvratil, i think i have see the bug in libical, can you check with my pull request ?

**Comment by danvratil** [comment]
I think your patch fixes the symptoms of the bug but not the root cause, which is how did you end up with both UNTIL and COUNT having a value set in the first place. I'm not a libical dev and don't know the internals very well, I think we'll just have to wait for @winterz to take look at this.

**Comment by ksmurchison** [comment]
Fixed by PR #391

**Comment by winterz** [comment]
will be included with the 3.0.6 release

**Comment by Vogtinator** [comment]
With 3.0.6 it seems to generate an empty RRULE instead of an invalid one now, which also leads to an invalid .ical file:

```
BEGIN:STANDARD
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
DTSTART:19971026T030000
RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
END:STANDARD
BEGIN:STANDARD
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
DTSTART:19800928T030000
RRULE:
RDATE:19950924T030000
END:STANDARD
```

**Comment by ksmurchison** [comment]
What set of API calls lead to the empty RRULE?

**Comment by Vogtinator** [comment]
I'm not sure, but I guess those here (@danvratil could probably confirm that, I'm familiar with neither ical nor kcalendarcore code):

https://github.com/KDE/kcalendarcore/blob/cf68c74439d72687fba2bbf65818af46982fb283/src/icaltimezones.cpp#L385-400

It sets both `r.count` and `r.until`, that might the the issue?

**Comment by ksmurchison** [comment]
So, what is the suggested behavior if an application goes out of its way to create an invalid RRULE?
I suppose we can make an arbitrary decision that if the RRULE contains both UNTIL and COUNT, that UNTIL takes precedence.

**Comment by Vogtinator** [comment]
That would match what the fix for kcalendarcore did - just removing the `r.count` assignment and keeping the `r.until` one.

**Comment by ksmurchison** [comment]
Fixed with the commit above

**Comment by Vogtinator** [comment]
Thanks!

**Comment by winterz** [comment]
I'll merge to the 3.0 branch so the fix will be included with the 3.0.8.  Hopefully coming soon.
